src/Microsoft.ML.PCA/PcaTransform.cs

```diff
@@ -98,40 +98,72 @@ public bool TryUnparse(StringBuilder sb)
             }
         }
 
+        public sealed class ColumnInfo
+        {
+            public readonly string Input;
+            public readonly string Output;
+            public readonly string WeightColumn;
+            public readonly int Rank;
+            public readonly int Oversampling;
+            public readonly bool Center;
+            public readonly int? Seed;
+
+            /// <summary>
+            /// Describes how the transformer handles one column pair.
+            /// </summary>
+            public ColumnInfo(string input,
+                              string output,
+                              string weightColumn = PcaEstimator.Defaults.WeightColumn,
+                              int rank = PcaEstimator.Defaults.Rank,
+                              int overSampling = PcaEstimator.Defaults.Oversampling,
+                              bool center = PcaEstimator.Defaults.Center,
+                              int? seed = null)
+            {
+                Input = input;
+                Output = output;
+                WeightColumn = weightColumn;
+                Rank = rank;
+                Oversampling = overSampling;
+                Center = center;
+                Seed = seed;
+                Contracts.CheckUserArg(Oversampling >= 0, nameof(Oversampling), "Oversampling must be non-negative.");
```

Comment:
>Contracts [](start = 16, length = 9)

Can you add 
`Contracts.CheckParam(0 < Rank && Rank <= Dimension, nameof(Rank), "Rank must be positive, and at most the dimension of untransformed data"); `
as well?
Also change it to CheckParam

Reply:
That check already exists in TransformInfo constructor. The Dimension information is kept in that class and is not available in ColumnInfo class, because it depends on the input schema.

---
In reply to: [227573435](https://github.com/dotnet/machinelearning/pull/1333#discussion_r227573435) [](ancestors = 227573435)

Reply:
Yes, it's part of TransformInfo constructor, but in same time we hit that constructor only during training stage.
You can create estimator, put it as part of chain, and never call it, or call it few screens later, and only after it hit TransformInfo it would throw exception.
If you put it in ColumnInfo as well, it will throw immediately.

---
In reply to: [227624838](https://github.com/dotnet/machinelearning/pull/1333#discussion_r227624838) [](ancestors = 227624838,227573435)

Reply:
The `Dimension` is equal to the vector size of the column, and it's only known at the time of `Fit`, not before. So there is no way to throw early.

---
In reply to: [227869834](https://github.com/dotnet/machinelearning/pull/1333#discussion_r227869834) [](ancestors = 227869834,227624838,227573435)

Reply:
At least you can check rank negativity.

---
In reply to: [227869834](https://github.com/dotnet/machinelearning/pull/1333#discussion_r227869834) [](ancestors = 227869834,227624838,227573435)

Reply:
True :)

---
In reply to: [227877885](https://github.com/dotnet/machinelearning/pull/1333#discussion_r227877885) [](ancestors = 227877885,227869834,227624838,227573435)

Reply:
Positivity

---
In reply to: [227879084](https://github.com/dotnet/machinelearning/pull/1333#discussion_r227879084) [](ancestors = 227879084,227877885,227869834,227624838,227573435)

Reply:
Added positivity check for rank

---
In reply to: [227879168](https://github.com/dotnet/machinelearning/pull/1333#discussion_r227879168) [](ancestors = 227879168,227879084,227877885,227869834,227624838,227573435)